src/udev/udevd.c

```diff
@@ -351,16 +351,42 @@ static void notify_ready(void) {
 }
 
 /* reload requested, HUP signal received, rules changed, builtin changed */
-static void manager_reload(Manager *manager) {
+static void manager_reload(Manager *manager, bool force) {
+        usec_t now_usec;
+        UdevRules *rules;
+        int r;
+
         assert(manager);
 
+        assert_se(sd_event_now(manager->event, CLOCK_MONOTONIC, &now_usec) >= 0);
+        if (!force && now_usec < usec_add(manager->last_usec, 3 * USEC_PER_SEC))
+                /* check for changed config, every 3 seconds at most */
+                return;
+        manager->last_usec = now_usec;
+
+        /* Reload SELinux label database, to make the child inherit the up-to-date database. */
+        mac_selinux_maybe_reload();
+
+        /* Nothing changed. It is not necessary to reload. */
+        if (!udev_rules_should_reload(manager->rules) && !udev_builtin_validate())
+                return;
```

Comment:
So, this is probably not going to be liked by people who build reproducible OS images, and things like ostree systems. For those the mtimes carry little value, as reproducing the very same image x times successfully would mean all file contents and all metadata will match -- except of for the mtime which is different eahc time. Hence for example ostree sets the mtimes all to zero (i.e. jan 1st, 1970) in their images, because they want to share inodes between images (via hardlink farms) and they only can do that if the mtimes carry no value. rules files are often stored in /usr/ and hence have 1970 timestamps.

So, I am sure that eating up explicit reload requests in ostree setups should continue to have an effect, as mtimes are useless there in /usr/ to detect changes.

to salvage that, two ideas:

1. honour the fact that ostree sets mtime to 0 for its inodes: i.e. don't suppress reloads if the dir mtime is 0.
2. add an env var that allows overriding the logic.

Reply:
Interesting. But I'd like to keep the current behavior. Reasons:
- If a package or os image is updated or replaced, then udevd should be restarted.
- I guess .rules or .link files under /etc or /run has non-zero mtime, and are managed by admin, instead of packager or image provider.
- The `stats_by_path` hashmap has full `struct stat`, not only mtime. So, of course that's not perfect, but when a .rules or .link file is changed, its file size is also changed in most cases, and we can detect the change even if mtime is not changed.

If the zero mtime introduces any real issues, then I'd like to address that in a later PR, as the current logic is already used by networkd (and IIRC, PID1 also use the logic partially).